components/registry-facade/pkg/handover/handover.go

```diff
@@ -0,0 +1,139 @@
+package handover
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+)
+
+// OfferHandover opens a Unix socket on socketFN and waits for another process to ask
+// for the listeners socket file descriptor. Once that happens, it closes the Unix socket and returns.
+// If the context is canceled before someone asks for the listener's socket,
+// this function returns context.Canceled.
+func OfferHandover(ctx context.Context, socketFN string, l *net.TCPListener) error {
+	skt, err := net.Listen("unix", socketFN)
+	if err != nil {
+		return xerrors.Errorf("cannot create handover socket: %w", err)
+	}
+	defer skt.Close()
+
+	done := make(chan struct{})
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		recv, err := skt.Accept()
+		if err != nil {
+			return xerrors.Errorf("cannot accept incoming handover connection: %w", err)
+		}
+		defer recv.Close()
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		listenConn := recv.(*net.UnixConn)
+		err = sendListener(listenConn, l)
+		if err != nil {
+			return xerrors.Errorf("cannot send listener: %w", err)
+		}
+
+		defer close(done)
+		return nil
+	})
+	eg.Go(func() error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-done:
+			return nil
+		}
+	})
+	return eg.Wait()
+}
+
+// ReceiveHandover requests a net.Listener handover from a Unix socket on socketFN.
+// If the context cancels before the transfer is complete, context.Canceled is returned.
+func ReceiveHandover(ctx context.Context, socketFN string) (l net.Listener, err error) {
+	conn, err := net.Dial("unix", socketFN)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	return receiveListener(ctx, conn.(*net.UnixConn))
+}
+
+// sendListener sends a copy of a TCP listener's file descriptor to a Unix socket.
+// Callers should close l upon successful return of this function.
+// Use in conjunction with receiveListener().
+func sendListener(conn *net.UnixConn, l *net.TCPListener) error {
+	connf, err := conn.File()
+	if err != nil {
+		return err
+	}
+	defer connf.Close()
+	sktfd := int(connf.Fd())
+
+	lf, err := l.File()
+	if err != nil {
+		return err
+	}
+	lfd := int(lf.Fd())
+
+	rights := unix.UnixRights(lfd)
```

Comment:
A comment that `unix.UnixRights` does encode the fd would be king. Had to look that up to understand what's going on.